File: src/main/webapp/assets/js/space/RelativePointsObject.js
Engine.define('RelativePointsObject', ['BaseObject', 'Geometry', 'Point'], function(){

    var Point = Engine.require('Point');
    var Geometry = Engine.require('Geometry');
    var BaseObject = Engine.require('BaseObject');

    function RelativePointsObject(x, y, angle, points) {
        BaseObject.apply(this, arguments);
    }
    RelativePointsObject.prototype = Object.create(BaseObject.prototype);
    RelativePointsObject.prototype.constructor = RelativePointsObject;

    RelativePointsObject.zero = new Point(0, 0);

    RelativePointsObject.prototype.getPoints = function(){
        var me = this;
        return this.points.map(function(v){
            var p = Geometry.translate(RelativePointsObject.zero, v, me.angle);
            if(p === v) {
                p = p.clone();
            }
            p.x += me.x;
            p.y += me.y;
            return p;
        });
    };

    return RelativePointsObject;
});

File: src/main/java/org/forweb/drift/entity/drift/RelativePointsObject.java
package org.forweb.drift.entity.drift;


import org.forweb.geometry.services.PointService;
import org.forweb.geometry.shapes.Point;

public class RelativePointsObject extends BaseObject{

    public static Point zero = new Point(0, 0);


    RelativePointsObject(double x, double y, double angle, Point[] points) {
        super(x, y, angle, points);
    }

    public Point[] getPoints() {
        Point[] points = super.getPoints();
        Point[] out = new Point[points.length];
        for(int i = 0; i < points.length; i++) {
            Point p = PointService.translate(zero, points[i], getAngle());
            out[i] = new Point(p.getX() + getX(), p.getY() + getY());
        }
        return out;
    }
}
